kratos/processes/apply_periodic_boundary_condition_process.cpp

```diff
@@ -149,92 +150,126 @@ template <int TDim>
-
-
-            if (TDim == 3) // TODO: This function can be optimized using template specialization. if constexpr will fail due to unused variables declared
+            #pragma omp critical
             {
-                auto constraint7 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_x, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,0), constant_z);
-                auto constraint8 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_y, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,1), constant_z);
-                auto constraint9 = r_clone_constraint.Create(++current_num_constraint, master_node, r_var_z, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,2), constant_z);
-
-                mrMasterModelPart.AddMasterSlaveConstraint(constraint7);
-                mrMasterModelPart.AddMasterSlaveConstraint(constraint8);
-                mrMasterModelPart.AddMasterSlaveConstraint(constraint9);
+                int current_num_constraint = mrMasterModelPart.GetRootModelPart().NumberOfMasterSlaveConstraints();
+                auto p_constraint1 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_x, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,0), constant_x);
+                auto p_constraint2 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_y, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,1), constant_x);
+                auto p_constraint3 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_z, rSlaveNode, r_var_x, master_weight * mTransformationMatrixVariable(0,2), constant_x);
+
+                auto p_constraint4 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_x, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,0), constant_y);
+                auto p_constraint5 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_y, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,1), constant_y);
+                auto p_constraint6 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_z, rSlaveNode, r_var_y, master_weight * mTransformationMatrixVariable(1,2), constant_y);
+
+                auto p_constraint7 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_x, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,0), constant_z);
+                auto p_constraint8 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_y, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,1), constant_z);
+                auto p_constraint9 = r_clone_constraint.Create(++current_num_constraint, r_master_node, r_var_z, rSlaveNode, r_var_z, master_weight * mTransformationMatrixVariable(2,2), constant_z);
```

Comment:
I'd do this in a nested loop to reduce copy-pasta. ```TDim``` is known at compile time so the loops would automatically get unrolled by the compiler anyway.